core/src/consensus/tendermint/stake/action_data.rs

```diff
@@ -79,10 +87,54 @@ impl<'a> StakeAccount<'a> {
     }
 }
 
+pub struct Stakeholders(BTreeSet<Address>);
+
+impl Stakeholders {
+    pub fn load_from_state(state: &TopLevelState) -> StakeResult<Stakeholders> {
+        let key = get_stakeholder_addresses_key();
+        let action_data = state.action_data(&key)?;
+
+        let mut addresses = BTreeSet::new();
+
+        if let Some(rlp) = action_data.as_ref().map(|x| UntrustedRlp::new(x)) {
+            for i in 0..rlp.item_count()? {
+                addresses.insert(rlp.val_at(i)?);
+            }
+        }
+
+        Ok(Stakeholders(addresses))
+    }
+
+    pub fn save_to_state(&self, state: &mut TopLevelState) -> StakeResult<()> {
+        let key = get_stakeholder_addresses_key();
+        let mut rlp = RlpStream::new();
+        rlp.begin_list(self.0.len());
+        for address in self.0.iter() {
+            rlp.append(address);
+        }
+        state.update_action_data(&key, rlp.drain().into_vec())?;
+        Ok(())
+    }
+
+    #[cfg(test)]
+    pub fn contains(&self, address: &Address) -> bool {
+        self.0.contains(address)
+    }
+
+    pub fn update(&mut self, account: &StakeAccount) {
+        if account.balance > 0 {
+            self.0.insert(*account.address);
+        } else {
+            self.0.remove(account.address);
+        }
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
     use cstate::tests::helpers;
+    use rand::{thread_rng, Rng};
```

Comment:
Please use [rand_xorshift](https://crates.io/crates/rand_xorshift) for tests.
It makes test reproducible because it's a deterministic random generator.
Please see https://github.com/CodeChain-io/codechain/commit/9c686e4474eccde51a5d2cf065d47d336dc80146